[mlir][Vector] Refactor VectorEmulateNarrowType.cpp

banach-space opened the pull request:
This is PR refactors `alignedConversionPrecondition` and adds new helper
hooks.

**Update `alignedConversionPrecondition` (1)**

This method doesn't require the vector type for the "container" argument. The
underlying element type is sufficient. The corresponding argument has been
renamed as `containerTy` - this is meant as the multi-byte container element
type (`i8`, `i16`, `i32`, etc). With this change, the updated invocations of
`alignedConversionPrecondition` (in e.g. `RewriteAlignedSubByteIntExt`) make it
clear that the container element type is assumed to be `i8`.

**Update alignedConversionPrecondition (2):**

The final check in `alignedConversionPrecondition` has been replaced with a new
helper method, `isSubByteVecFittable`. This helper hook is now also re-used in
`ConvertVectorTransferRead` (to improve code re-use).

**Other updates**

Extended + unified comments.

**Implements**: https://github.com/llvm/llvm-project/issues/123630

banach-space committed:
[mlir][Vector] Update VectorEmulateNarrowType.cpp (4/N)

This is PR 4 in a series of N patches aimed at improving
"VectorEmulateNarrowType.cpp". This is mainly minor refactoring, no
major functional changes are made/added.

1. Update `alignedConversionPrecondition` (1):

This method didn't require the vector type for the "destination"
argument. The underlying element type is sufficient. The corresponding
argument has been renamed as `multiByteScalarTy` - this is meant as the
multi-byte emulated type (`i8`, `i16`, `i32`, etc).

2. Update `alignedConversionPrecondition` (2):

In #121298, we replaced `dstElemBitwidt` in this calculation:

```cpp
  const int numSrcElemsPerDestElem = dstElemBitwidth / srcElemBitwidth;
```

with the hard-coded value of 8:
```cpp
  const int numSrcElemsPerDestElem = 8 / srcElemBitwidth;
```

That was correct as for the patterns for which this hook was/is used:

  * `RewriteAlignedSubByteIntExt`,
  * `RewriteAlignedSubByteIntTrunc`.

The destination type (or, more precisely, the emulated type) was always
`i8`.

In this PR, I am switching back to a more generic approach - the
calculation should take into account the bit-width of the emulated type.

Note that at the call sites I am passing `i8` as the emulated type, so the
end-result is effectively identical. However, the intent is clearer, i.e.,
the underlying value is 8 because the emulated type happens to be `i8`
(as opposed using a magic number).

3. Update alignedConversionPrecondition (3):

The final check has been replaced with a new helper method,
`isSubByteVecFittable`. This new method is also re-used within the code
and hopefully will allow us more code re-use moving forward (to avoid
re-implementing the same condition).

NEXT STEPS (1):

We need to clarify the meaning of "source" and "destination" types.
Currently the usage is ambiguous.

For example, for this `arith.extsi` Op, `vector<8xi2>` and
`vector<8xi32>` are the "source" and "destination" types, respectively:

```mlir
  %0 = arith.extsi %arg0 : vector<8xi2> to vector<8xi32>
}
```

However, patterns like `RewriteAlignedSubByteIntExt` introduce
`vector.bitcast` Ops like this:

```mlir
  %bitcast = vector.bitcast %arg0 : vector<8xi2> to vector<2xi8>
```

I've noticed that we tend to mix `vector<2xi8>` and `vector<8xi32>` as
the destination types and that should be clarified.

NEXT STEPS (2):

With this PR, I am introducing explicit references to "sub-byte" as
that is effectively what this logic is used of (i.e. for emulating
"sub-byte" types). We should either generalise (which would include
increasing test coverage) or restrict everything to "sub-byte" type
emulation.

dcaballe reviewed (commented):
Merge with previous, please

banach-space commented:
> Merge with previous, please

Currently as a draft:

* https://github.com/llvm/llvm-project/pull/123633

GitHub issue for more context: https://github.com/llvm/llvm-project/issues/123630

banach-space commented:
CC @ziereis

banach-space commented:
**UPDATE 6/2/25**

As all dependencies have been merged, I've re-based this on top of `main`.

banach-space commented:
**UPDATE 1/3/25**

All dependencies have been merged, this is now ready for review.

banach-space commented:
Ping @lialan , this is the refactor that we discussed in https://github.com/llvm/llvm-project/pull/115922

Thanks!

banach-space commented:
**UPDATE 8/3/25**

Simplified the summary. Apologies, originally I packed too much info there.

lialan commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -1127,7 +1159,8 @@ struct ConvertVectorTransferRead final
 
     // Note, per-element-alignment was already verified above.
-    bool isFullyAligned = origElements % emulatedPerContainerElem == 0;
+    bool isFullyAligned =
```
I wonder if we should also change the name of `isFullyAligned`

banach-space replied:
Are you suggesting to rename it everywhere? Note that ATM this name is quite widely used throughout this file:
```bash 
$ rg "isFullyAligned" mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp | wc -l
      14
```

I am fine renaming it, though I will ask you for a suggestion 😅 

Perhaps let me start by explaining the rationale for this name. Basically, most patterns implement this high level logic (see e.g. [here](https://github.com/llvm/llvm-project/blob/5ce4045384d7c2544185b0dbcb6222d06beb47dc/mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp#L499-L504)):

```cpp
    // Check per-element alignment.
    if (containerBits % emulatedBits != 0) {
      return rewriter.notifyMatchFailure(
          op, "impossible to pack emulated elements into container elements "
              "(bit-wise misalignment)");
    }
    
    // (...) Some code here
    
    // Note, per-element-alignment was already verified above.
    bool isFullyAligned = origElements % emulatedPerContainerElem == 0;
```

As a concrete example:
* `vector<3x2xi2>` is "per-element" aligned with `vector<2xi8>` (because we can fit exactly 4 x `i2` into `i8`)
* `vector<3x2xi2>` is not full aligned with `vector<2xi8>` as (3 x 2 = ) **6 % 4** (= 8 / 2) **!= 0**

Let me know if this makes sense. If not, we can iterate :)

lialan replied:
Naming is indeed hard! I asked my AI and I got `isDivisibleInSize` out from a bunch of suggestions. :-)

lialan replied:
I actually prefer option 3..... Once `isPerElementAlligned` is tested we don't need subsequent references to it.

It also reduces mental burden while reading the code.

banach-space replied:
Updated in this [commit](https://github.com/llvm/llvm-project/pull/123529/commits/3abe26ca6a3c7579190e91be807e1d07ecc2c234)

lialan commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -1428,41 +1461,76 @@ LogicalResult BitCastRewriter::commonPrecondition(PatternRewriter &rewriter,
-    return rewriter.notifyMatchFailure(op, "unaligned cases are not supported");
-  if (srcElemBitwidth != 2 && srcElemBitwidth != 4)
+  // TODO: This is validating the inputs rather than checking the conditions
+  // documented above. Replace with an assert.
```
Should this be an assert instead of returning failure?

banach-space replied:
Done

lialan commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -1428,41 +1461,76 @@ LogicalResult BitCastRewriter::commonPrecondition(PatternRewriter &rewriter,
+
+  // TODO: Remove this condition - the assert above (and
+  // commonConversionPrecondtion) takes care of that.
+  if (multiByteBits < 8)
```
if `multiByteBits % 8 == 0` then this cannot happen, unless it is zero.

banach-space replied:
Indeed, let me remove it, thanks!

lialan commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -1428,41 +1461,76 @@ LogicalResult BitCastRewriter::commonPrecondition(PatternRewriter &rewriter,
+  if (subByteBits != 2 && subByteBits != 4)
     return rewriter.notifyMatchFailure(
-        op, "only src bitwidth of 2 or 4 is supported at this moment");
+        op, "only 2-bit and 4-bit sub-byte type is supported at this moment");
```
1 bit?

lialan replied:
I know some downstream projects try to use 1bit, and I think upstream shouldn't trivially block it in this way. They can contribute i1 tests for sure but overall the code here should support 1-bit scenarios without problem.

lialan commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -1091,6 +1091,38 @@ struct ConvertVectorMaskedLoad final
+///   * vector<4xi4> -> i8 - yes (N = 2)
+///   * vector<3xi4> -> i8 - no (N would have to be 1.5)
+///   * vector<3xi2> -> i16 - no (N would have to be 0.5)
+static bool isSubByteVecFittable(VectorType subByteVecTy,
```
suggestion use a clearer name: `sizeFitsInMultiByteScalarTy` or something else. 

While I was reading the invocation of this function somewhere down this file, I felt this function's name a bit confusing.

banach-space replied:
Naming is hard! Thanks for the suggestion. How about: `fitsInMultiByteContainerTy`:
* dropped `size` as we are not passing `size`, and 
* replaced `Scalar` with `Container` to solidify the intent (while it is a "scalar", it's a special scalar that is meant to be a container 😅 )

WDYT? Please give me your honest answer, it's important to make this clear for everyone.

lialan replied:
`fitsInMultiByteContainerTy` works!

lialan commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -1428,41 +1461,76 @@ LogicalResult BitCastRewriter::commonPrecondition(PatternRewriter &rewriter,
-        op, "only src bitwidth of 2 or 4 is supported at this moment");
+        op, "only 2-bit and 4-bit sub-byte type is supported at this moment");
+
+  // Condition 1.
```
Probably don't need Condition 1, condition 2, if you we do not want to comment on what the conditions are.

banach-space replied:
I actually did intend to add comments, these are the "per-element" and "full" alignment conditions I mentioned elsewhere. The idea was to highlight that it's the same conditions checked in various places. Let me add comments.

Thanks!

lialan commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -1964,8 +2033,9 @@ struct RewriteAlignedSubByteIntTrunc : OpRewritePattern<arith::TruncIOp> {
     // to reuse the existing precondition logic.
-    if (failed(alignedConversionPrecondition(rewriter, dstVecType, srcVecType,
-                                             truncOp)))
+    Type containerType = rewriter.getI8Type();
```
Since it is only 1 use, can fold it into subsequent line.

lialan commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -1899,8 +1967,9 @@ struct RewriteAlignedSubByteIntExt : OpRewritePattern<ConversionOpType> {
     // Check general alignment preconditions.
-    if (failed(alignedConversionPrecondition(rewriter, srcVecType, dstVecType,
-                                             conversionOp)))
+    Type containerType = rewriter.getI8Type();
```
Since it is only 1 use, can fold it into subsequent line.

banach-space committed:
fixup! [mlir][Vector] Update VectorEmulateNarrowType.cpp (4/N)

Address comments from Alan

lialan reviewed (approved):
LGTM, let's wrap up the remaining unresolved conversations.

banach-space committed:
fixup! fixup! [mlir][Vector] Update VectorEmulateNarrowType.cpp (4/N)

isFullyAligned -> isDivisibleInSize

banach-space committed:
fixup! fixup! fixup! [mlir][Vector] Update VectorEmulateNarrowType.cpp (4/N)

Add minor missing re-naming (otherwise there are inconsistent names remaining))
